bpf/http_sock.c

```diff
@@ -0,0 +1,194 @@
+// we don't have the full picture for the socket.
+// 
+// Note: A current limitation is that likely we won't capture the first accept request. The
+// process may have already reached accept, before the instrumenter has launched.
+SEC("kretprobe/sys_accept4")
+int BPF_KRETPROBE(kretprobe_sys_accept4, uint fd)
+{
+    u64 id = bpf_get_current_pid_tgid();
+
+    if (!valid_pid(id)) {
+        return 0;
+    }
+
+    bpf_dbg_printk("=== accept 4 ret id=%d ===", id);
+
+    // The file descriptor is the value returned from the accept4 syscall.
+    // If we got a negative file descriptor we don't have a connection
+    if ((int)fd < 0) {
+        goto cleanup;
+    }
+
+    sock_args_t *args = bpf_map_lookup_elem(&active_accept_args, &id);
+    if (!args) {
+        bpf_dbg_printk("No sock info %d", id);
+        goto cleanup;
+    }
+
+    http_connection_info_t info = {};
+
+    parse_accept_socket_info(args, &info);
+    sort_connection_info(&info);
+    dbg_print_http_connection_info(&info);
+
+    bpf_map_update_elem(&filtered_connections, &info, &id, BPF_ANY); // On purpose BPF_ANY, we want to overwrite stale
+
+cleanup:
+    bpf_map_delete_elem(&active_accept_args, &id);
+    return 0;
+}
+
+// Used by connect so that we can grab the sock details
+SEC("kprobe/tcp_connect")
+int BPF_KPROBE(kprobe_tcp_connect, struct sock *sk) {
+    u64 id = bpf_get_current_pid_tgid();
+
+    if (!valid_pid(id)) {
+        return 0;
+    }
+
+    bpf_dbg_printk("=== tcp connect %llx ===", id);
+
+    u64 addr = (u64)sk;
+
+    sock_args_t args = {};
+
+    args.addr = addr;
+    args.accept_time = bpf_ktime_get_ns();
+
+    bpf_map_update_elem(&active_connect_args, &id, &args, BPF_ANY);
+
+    return 0;
+}
+
+// We tap into sys_connect so we can track properly the processes doing
+// HTTP client calls
+SEC("kretprobe/sys_connect")
+int BPF_KRETPROBE(kretprobe_sys_connect, int fd)
+{
+    u64 id = bpf_get_current_pid_tgid();
+
+    if (!valid_pid(id)) {
+        return 0;
+    }
+
+    bpf_dbg_printk("=== connect ret id=%d, pid=%d ===", id, pid_from_pid_tgid(id));
+
+    // The file descriptor is the value returned from the connect syscall.
+    // If we got a negative file descriptor we don't have a connection, unless we are in progress
+    if (fd < 0 && (fd != -EINPROGRESS)) {
+        goto cleanup;
+    }
+
+    sock_args_t *args = bpf_map_lookup_elem(&active_connect_args, &id);
+    if (!args) {
+        bpf_dbg_printk("No sock info %d", id);
+        goto cleanup;
+    }
+
+    http_connection_info_t info = {};
+
+    parse_connect_sock_info(args, &info);
+    sort_connection_info(&info);
+    dbg_print_http_connection_info(&info);
+
+    bpf_map_update_elem(&filtered_connections, &info, &id, BPF_ANY); // On purpose BPF_ANY, we want to overwrite stale
+
+cleanup:
+    bpf_map_delete_elem(&active_connect_args, &id);
+    return 0;
+}
+
+SEC("kprobe/sys_exit")
+int BPF_KPROBE(kprobe_sys_exit, int status) {
```

Comment:
I track sys_exit here to capture the program name of an ending program, so that we can find it later when producing the events for system wide instrumentation.